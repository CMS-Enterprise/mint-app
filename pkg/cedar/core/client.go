package cedarcore

import (
	"context"
	"net/http"
	"time"

	ld "gopkg.in/launchdarkly/go-server-sdk.v5"

	"github.com/go-openapi/runtime"
	httptransport "github.com/go-openapi/runtime/client"
	"github.com/go-openapi/strfmt"
	cache "github.com/patrickmn/go-cache"
	"go.uber.org/zap"

	"github.com/cmsgov/easi-app/pkg/appcontext"
	apiclient "github.com/cmsgov/easi-app/pkg/cedar/core/gen/client"
	"github.com/cmsgov/easi-app/pkg/flags"
)

const (
	cedarCoreEnabledKey     = "cedarCoreEnabled"
	cedarCoreEnabledDefault = false
)

// NewClient builds the type that holds a connection to the CEDAR Core API
func NewClient(ctx context.Context, cedarHost string, cedarAPIKey string, ldClient *ld.LDClient) *Client {
	fnEmit := func(ctx context.Context) bool {
		lduser := flags.Principal(ctx)
		result, err := ldClient.BoolVariation(cedarCoreEnabledKey, lduser, cedarCoreEnabledDefault)
		if err != nil {
			appcontext.ZLogger(ctx).Info(
				"problem evaluating feature flag",
				zap.Error(err),
				zap.String("flagName", cedarCoreEnabledKey),
				zap.Bool("flagDefault", cedarCoreEnabledDefault),
				zap.Bool("flagResult", result),
			)
		}
		return result
	}

	c := cache.New(cache.NoExpiration, cache.NoExpiration) // Don't expire data _or_ clean it up

	hc := http.DefaultClient

	client := &Client{
		cedarCoreEnabled: fnEmit,
		auth: httptransport.APIKeyAuth(
			"x-Gateway-APIKey",
			"header",
			cedarAPIKey,
		),
		sdk: apiclient.New(
			httptransport.New(
				cedarHost,
				apiclient.DefaultBasePath,
				apiclient.DefaultSchemes,
			),
			strfmt.Default,
		),
		hc:    hc,
		cache: c,
	}

	// Start cache refresh for systems
	client.startCacheRefresh(ctx, time.Minute*5, client.populateSystemSummaryCache)

	return client
}

// Client represents a connection to the CEDAR Core API
type Client struct {
	cedarCoreEnabled func(context.Context) bool
	auth             runtime.ClientAuthInfoWriter
	sdk              *apiclient.CEDARCoreAPI
	hc               *http.Client
	cache            *cache.Cache
}

// startCacheRefresh starts a goroutine that will run `populateCache` based on cacheRefreshTime
// This function returns no errors, and only logs when something goes wrong
func (c *Client) startCacheRefresh(ctx context.Context, cacheRefreshTime time.Duration, populateCache func(context.Context) error) {
	ticker := time.NewTicker(cacheRefreshTime)
	go func(ctx context.Context) {
		for {
			err := populateCache(ctx)
			if err != nil {
				appcontext.ZLogger(ctx).Error("Failed to refresh cache", zap.Error(err))
			}
			// Wait for the ticker. This will block the current goroutine until the ticker sends a message over the channel
			<-ticker.C
		}
	}(ctx)
}
