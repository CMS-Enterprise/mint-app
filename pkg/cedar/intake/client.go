// Package intake serves as an Anti-Corruption Layer (in Domain Driven Design
// parlance), keeping the implementation details of working with the
// downstream CEDAR Intake API well separated, so the rest of the internal
// workings of the EASi code are not using anything defined by the
// external API
package intake

import (
	"context"
	"fmt"
	"net/http"
	"strconv"

	"github.com/go-openapi/runtime"
	httptransport "github.com/go-openapi/runtime/client"
	"github.com/go-openapi/strfmt"
	"go.uber.org/zap"
	ld "gopkg.in/launchdarkly/go-server-sdk.v5"

	"github.com/cmsgov/easi-app/pkg/appcontext"
	apiclient "github.com/cmsgov/easi-app/pkg/cedar/intake/gen/client"
	"github.com/cmsgov/easi-app/pkg/cedar/intake/gen/client/health_check"
	"github.com/cmsgov/easi-app/pkg/cedar/intake/gen/client/intake"
	"github.com/cmsgov/easi-app/pkg/cedar/intake/translation"
	"github.com/cmsgov/easi-app/pkg/flags"
	"github.com/cmsgov/easi-app/pkg/models"
)

const (
	emitFlagKey = "emit-to-cedar"
	emitDefault = false
)

// NewClient builds the type that holds a connection to the CEDAR Intake API
func NewClient(cedarHost string, cedarAPIKey string, ldClient *ld.LDClient) *Client {
	fnEmit := func(ctx context.Context) bool {
		// this is the conditional way of stopping us from submitting to CEDAR; see EASI-1025
		lduser := flags.Principal(ctx)
		result, err := ldClient.BoolVariation(emitFlagKey, lduser, emitDefault)
		if err != nil {
			appcontext.ZLogger(ctx).Info(
				"problem evaluating feature flag",
				zap.Error(err),
				zap.String("flagName", emitFlagKey),
				zap.Bool("flagDefault", emitDefault),
				zap.Bool("flagResult", result),
			)
		}
		return result
	}

	hc := http.DefaultClient

	return &Client{
		emitToCedar: fnEmit,
		auth: httptransport.APIKeyAuth(
			"x-Gateway-APIKey",
			"header",
			cedarAPIKey,
		),
		sdk: apiclient.New(
			httptransport.New(
				cedarHost,
				apiclient.DefaultBasePath,
				apiclient.DefaultSchemes,
			),
			strfmt.Default,
		),
		hc: hc,
	}
}

// Client represents a connection to the CEDAR Intake API
type Client struct {
	emitToCedar func(context.Context) bool
	auth        runtime.ClientAuthInfoWriter
	sdk         *apiclient.CEDARIntake
	hc          *http.Client
}

// CheckConnection hits the CEDAR Intake API `/healthcheck` endpoint to verify
// that our connection is functional
func (c *Client) CheckConnection(ctx context.Context) error {
	if !c.emitToCedar(ctx) {
		return nil
	}

	params := health_check.NewHealthCheckParamsWithContext(ctx)
	params.HTTPClient = c.hc

	resp, err := c.sdk.HealthCheck.HealthCheck(params, c.auth)
	if err != nil {
		return err
	}
	if resp.Payload == nil {
		return fmt.Errorf("no response body received")
	}
	return nil
}

// PublishSystemIntake sends a system intake to CEDAR through the Intake API for eventual storage in Alfabet
func (c *Client) PublishSystemIntake(ctx context.Context, si models.SystemIntake) error {
	intakeObject := translation.TranslatableSystemIntake(si)
	return c.publishIntakeObject(ctx, &intakeObject)
}

// PublishBusinessCase sends a business case to CEDAR through the Intake API for eventual storage in Alfabet
func (c *Client) PublishBusinessCase(ctx context.Context, bc models.BusinessCase) error {
	intakeObject := translation.TranslatableBusinessCase(bc)
	return c.publishIntakeObject(ctx, &intakeObject)
}

// PublishGRTFeedback sends an item of GRT feedback to CEDAR through the Intake API for eventual storage in Alfabet
func (c *Client) PublishGRTFeedback(ctx context.Context, feedback models.GRTFeedback) error {
	intakeObject := translation.TranslatableFeedback(feedback)
	return c.publishIntakeObject(ctx, &intakeObject)
}

// PublishAction sends an action to CEDAR through the Intake API for eventual storage in Alfabet
func (c *Client) PublishAction(ctx context.Context, action models.Action) error {
	intakeObject := translation.TranslatableAction(action)
	return c.publishIntakeObject(ctx, &intakeObject)
}

// PublishNote sends a note to CEDAR through the Intake API for eventual storage in Alfabet
func (c *Client) PublishNote(ctx context.Context, note models.Note) error {
	intakeObject := translation.TranslatableNote(note)
	return c.publishIntakeObject(ctx, &intakeObject)
}

// private method for publishing anything that satisfies the translation.IntakeObject interface to CEDAR through the Intake API
func (c *Client) publishIntakeObject(ctx context.Context, model translation.IntakeObject) error {
	// constant values for now; may become non-constant if/when we revisit handling CEDAR validation errors

	// from the Swagger (cedar_intake.json, definitions/IntakeInput/properties/version):
	// "The version associated with the object in the body. This value can be incremented in the event a transaction needs to be resubmitted."
	const objectVersion = 1

	// from the Swagger (cedar_intake.json, paths/"/intake"/post/parameters/validatePayload):
	// "Determines if schema validation of the payload is performed syncronously before persisting the record or asyncronously after the record has been persisted"
	const isValidatedSynchronously = false // false for V1 schemas; this may change when we implement V2 schemas for EASI-1614

	id := model.ObjectID()
	objectType := model.ObjectType()

	if !c.emitToCedar(ctx) {
		appcontext.ZLogger(ctx).Info("snapshot publishing disabled", zap.String("object ID", id), zap.String("object type", objectType))
		return nil
	}

	input, err := model.CreateIntakeModel()

	if err != nil {
		appcontext.ZLogger(ctx).Error(
			"problem building cedar payload",
			zap.String("object ID", id),
			zap.String("object type", objectType),
			zap.Error(err),
		)
		return nil
	}

	params := intake.NewIntakeAddParamsWithContext(ctx)
	params.HTTPClient = c.hc
	params.Body = input

	params.ValidatePayload = strconv.FormatBool(isValidatedSynchronously)

	objectVersionStr := strconv.FormatInt(objectVersion, 10)
	params.Body.Version = &objectVersionStr

	resp, err := c.sdk.Intake.IntakeAdd(params, c.auth)
	if err != nil {
		return err
	}

	// TODO: we may need to read through the results for any "duplicate entry" error codes
	// that CEDAR _may_ send us, especially if we are re-transmitting entities... We
	// don't really care if there are dupes on their side, because we do NOT want to hold
	// any state on our side about who/what/when may have transmitted a previous copy
	// of the info to CEDAR.
	// appcontext.ZLogger(ctx).Info(fmt.Sprintf("results: %v", resp.Payload))
	_ = resp

	return nil
}
