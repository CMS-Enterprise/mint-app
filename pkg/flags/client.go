package flags

import (
	"context"
	"crypto/sha256"
	"fmt"
	"time"

	"github.com/cmsgov/mint-app/pkg/appcontext"

	"gopkg.in/launchdarkly/go-sdk-common.v2/lduser"
	ld "gopkg.in/launchdarkly/go-server-sdk.v5"
	"gopkg.in/launchdarkly/go-server-sdk.v5/ldfiledata"
	"gopkg.in/launchdarkly/go-server-sdk.v5/ldfilewatch"

	"github.com/cmsgov/mint-app/pkg/appconfig"
)

// DowngradeAssessmentTeamKey is the Flag Key in LaunchDarkly for downgrading assessment team users
const DowngradeAssessmentTeamKey = "downgradeAssessmentTeam"

// Config has all the parts for creating a new LD Client
type Config struct {
	Source         appconfig.FlagSourceOption
	Key            string
	Timeout        time.Duration
	FlagValuesFile string
}

// NewLaunchDarklyClient returns a client backed by Launch Darkly
func NewLaunchDarklyClient(config Config) (*ld.LDClient, error) {
	if config.Source == appconfig.FlagSourceLaunchDarkly {
		return ld.MakeClient(config.Key, config.Timeout)
	}

	defaultTimeout := 5 * time.Second

	if config.Source == appconfig.FlagSourceFile {

		ldConfig := ld.Config{
			DataSource: ldfiledata.DataSource().FilePaths(config.FlagValuesFile).Reloader(ldfilewatch.WatchFiles),
		}
		return ld.MakeCustomClient("fake_key", ldConfig, defaultTimeout)
	}

	// we default to an OFFLINE client for non-deployed environments
	return ld.MakeCustomClient("fake_offline_key", ld.Config{Offline: true}, 5*time.Second)
}

// Principal builds the LaunchDarkly user object for the
// currently authenticated principal.
func Principal(ctx context.Context) lduser.User {
	p := appcontext.Principal(ctx)
	key := UserKeyForID(p.ID())

	// this is a bit of a loose inference, assuming a user w/o Job Codes
	// is an Anonymous user. Over time, may want to consider adding
	// a `func Anonymous() bool` accessor to the authorizaion.Principal interface
	// definition instead of doing this inference
	authed := (p.AllowUSER() || p.AllowASSESSMENT())

	return lduser.
		NewUserBuilder(key).
		Anonymous(!authed).
		Build()
}

// UserKeyForID generates a user key from an ID
// we should not be using bare EUA IDs as identifiers to
// LaunchDarkly (per Jimil/ISSO), so we use a cryptographically
// secure one-way hash of the EUA ID as "key" for the LD User object.
func UserKeyForID(id string) string {
	h := sha256.New()
	_, _ = h.Write([]byte(id))
	return fmt.Sprintf("%x", h.Sum(nil))
}
