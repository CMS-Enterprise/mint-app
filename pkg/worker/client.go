package worker

import (
	"fmt"

	faktory "github.com/contribsys/faktory/client"
)

// NewClient returns a new faktory Client.
func NewClient() (*faktory.Client, error) {
	cl, err := faktory.Open()
	if err != nil {
		return nil, fmt.Errorf("error opening faktory client: %w", err)
	}
	return cl, nil
}
