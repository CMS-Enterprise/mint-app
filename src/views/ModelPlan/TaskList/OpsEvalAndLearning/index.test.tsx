import React from 'react';
import { MemoryRouter, Route } from 'react-router-dom';
import { MockedProvider } from '@apollo/client/testing';
import { render, screen, waitFor } from '@testing-library/react';
import {
  CcmInvolvmentType,
  GetOpsEvalAndLearningDocument,
  GetOpsEvalAndLearningQuery
} from 'gql/gen/graphql';

import {
  isCCWInvolvement,
  OpsEvalAndLearningContent,
  renderCurrentPage,
  renderTotalPages
} from './index';

type GetOpsEvalAndLearningType = GetOpsEvalAndLearningQuery['modelPlan']['opsEvalAndLearning'];

const opsEvalAndLearningMockData: GetOpsEvalAndLearningType = {
  __typename: 'PlanOpsEvalAndLearning',
  id: '123',
  ccmInvolvment: [],
  dataNeededForMonitoring: [],
  stakeholders: [],
  stakeholdersOther: '',
  stakeholdersNote: '',
  helpdeskUse: null,
  helpdeskUseNote: '',
  contractorSupport: [],
  contractorSupportOther: '',
  contractorSupportHow: 'Differ text',
  contractorSupportNote: '',
  iddocSupport: null,
  iddocSupportNote: ''
};

const opsEvalAndLearningMock = [
  {
    request: {
      query: GetOpsEvalAndLearningDocument,
      variables: { id: 'ce3405a0-3399-4e3a-88d7-3cfc613d2905' }
    },
    result: {
      data: {
        modelPlan: {
          id: 'ce3405a0-3399-4e3a-88d7-3cfc613d2905',
          modelName: 'My excellent plan that I just initiated',
          opsEvalAndLearning: opsEvalAndLearningMockData,
          operationalNeeds: [
            {
              __typename: 'OperationalNeed',
              id: '424213',
              modifiedDts: null
            }
          ]
        }
      }
    }
  }
];

describe('Model Plan Ops Eval and Learning', () => {
  it('computes total pages', () => {
    const totalPages9 = renderTotalPages(
      true,
      isCCWInvolvement([CcmInvolvmentType.YES_EVALUATION])
    );
    expect(totalPages9).toEqual(9);

    const totalPages6 = renderTotalPages(
      false,
      isCCWInvolvement([CcmInvolvmentType.YES_EVALUATION])
    );
    expect(totalPages6).toEqual(6);

    const totalPages8 = renderTotalPages(true, isCCWInvolvement([]));
    expect(totalPages8).toEqual(8);
  });

  it('computes current pages', () => {
    const currentPage = renderCurrentPage(
      1,
      true,
      isCCWInvolvement([CcmInvolvmentType.YES_EVALUATION])
    );
    expect(currentPage).toEqual(1);

    const currentPage2 = renderCurrentPage(
      5,
      false,
      isCCWInvolvement([CcmInvolvmentType.YES_EVALUATION])
    );
    expect(currentPage2).toEqual(2);
  });

  it('computes cmmi involvement', () => {
    const cmmiInvolvment = isCCWInvolvement([CcmInvolvmentType.YES_EVALUATION]);
    expect(cmmiInvolvment).toEqual(true);

    const cmmiInvolvmentFalse = isCCWInvolvement([]);
    expect(cmmiInvolvmentFalse).toEqual(false);
  });

  it('renders without errors', async () => {
    render(
      <MemoryRouter
        initialEntries={[
          '/models/ce3405a0-3399-4e3a-88d7-3cfc613d2905/task-list/ops-eval-and-learning'
        ]}
      >
        <MockedProvider mocks={opsEvalAndLearningMock} addTypename={false}>
          <Route path="/models/:modelID/task-list/ops-eval-and-learning">
            <OpsEvalAndLearningContent />
          </Route>
        </MockedProvider>
      </MemoryRouter>
    );

    await waitFor(() => {
      expect(
        screen.getByTestId('ops-eval-and-learning-form')
      ).toBeInTheDocument();
    });

    await waitFor(() => {
      expect(
        screen.getByTestId('ops-eval-and-learning-contractor-support-how')
      ).toHaveValue('Differ text');
    });
  });

  it('matches snapshot', async () => {
    const { asFragment } = render(
      <MemoryRouter
        initialEntries={[
          '/models/ce3405a0-3399-4e3a-88d7-3cfc613d2905/task-list/ops-eval-and-learning'
        ]}
      >
        <MockedProvider mocks={opsEvalAndLearningMock} addTypename={false}>
          <Route path="/models/:modelID/task-list/ops-eval-and-learning">
            <OpsEvalAndLearningContent />
          </Route>
        </MockedProvider>
      </MemoryRouter>
    );

    await waitFor(() => {
      expect(
        screen.getByTestId('ops-eval-and-learning-contractor-support-how')
      ).toHaveValue('Differ text');
    });
    expect(asFragment()).toMatchSnapshot();
  });
});
