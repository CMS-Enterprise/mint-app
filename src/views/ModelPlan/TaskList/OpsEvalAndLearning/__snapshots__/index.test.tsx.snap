// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Model Plan Ops Eval and Learning > matches snapshot 1`] = `
<DocumentFragment>
  <nav
    aria-label="Breadcrumbs"
    class="usa-breadcrumb usa-breadcrumb--wrap"
  >
    <ol
      class="usa-breadcrumb__list"
    >
      <li
        class="usa-breadcrumb__list-item"
      >
        <a
          class="usa-link usa-breadcrumb__link"
          href="/"
        >
          <span>
            Home
          </span>
        </a>
      </li>
      <li
        class="usa-breadcrumb__list-item"
      >
        <a
          class="usa-link usa-breadcrumb__link"
          href="/models/ce3405a0-3399-4e3a-88d7-3cfc613d2905/collaboration-area"
        >
          <span>
            Model collaboration area
          </span>
        </a>
      </li>
      <li
        class="usa-breadcrumb__list-item"
      >
        <a
          class="usa-link usa-breadcrumb__link"
          href="/models/ce3405a0-3399-4e3a-88d7-3cfc613d2905/collaboration-area/task-list"
        >
          <span>
            Model Plan task list
          </span>
        </a>
      </li>
      <li
        aria-current="page"
        class="usa-breadcrumb__list-item usa-current"
      >
        <span>
          Operations, evaluation, and learning
        </span>
      </li>
    </ol>
  </nav>
  <h1
    aria-live="polite"
    class="mint-h1 margin-top-4 margin-bottom-2"
    tabindex="-1"
  >
    Operations, evaluation, and learning
  </h1>
  <p
    class="margin-top-0 margin-bottom-1 font-body-lg"
    data-testid="model-plan-name"
  >
    for My excellent plan that I just initiated
  </p>
  <p
    class="margin-bottom-2 font-body-md line-height-sans-4"
  >
    If there's a question or field that is not applicable to your model or you don't currently know the answer, you may leave it blank. If you need help, ask a question using the link below.
  </p>
  <div
    class="padding-2 bg-primary-lighter"
  >
    <div
      class="display-flex"
      data-testid="ask-a-question"
    >
      <svg
        class="usa-icon text-primary margin-right-1"
        focusable="false"
        height="1em"
        role="img"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M20 2H4c-1.1 0-1.99.9-1.99 2L2 22l4-4h14c1.1 0 2-.9 2-2V4c0-1.1-.9-2-2-2zm-7 9h-2V5h2v6zm0 4h-2v-2h2v2z"
        />
      </svg>
      <button
        class="usa-button usa-button--unstyled"
        data-testid="ask-a-question-button"
        type="button"
      >
        Start a discussion
      </button>
    </div>
  </div>
  <form
    action="#"
    class="desktop:grid-col-6 margin-top-6"
    data-testid="ops-eval-and-learning-form"
  >
    <fieldset
      class="usa-fieldset"
      data-testid="fieldset"
    >
      <div
        class="usa-form-group margin-top-4"
        data-scroll="ops-eval-and-learning-stakeholders"
      >
        <label
          class="usa-label"
          data-testid="label"
          for="ops-eval-and-learning-stakeholders"
          id="label-ops-eval-and-learning-stakeholders"
        >
          What stakeholders do you plan to communicate with?
        </label>
        <div>
          <div
            class="easi-multiselect usa-combo-box margin-top-1 css-b62m3t-container"
            id="ops-eval-and-learning-stakeholders"
          >
            <span
              class="css-1f43avz-a11yText-A11yText"
              id="react-select-3-live-region"
            />
            <span
              aria-atomic="false"
              aria-live="polite"
              aria-relevant="additions text"
              class="css-1f43avz-a11yText-A11yText"
              role="log"
            />
            <div
              class=" css-1vl89cy-control"
            >
              <div
                class=" css-1fdsijx-ValueContainer"
              >
                <div
                  class=" css-1jqq78o-placeholder"
                  id="react-select-3-placeholder"
                >
                  0 selected
                </div>
                <div
                  class=" css-qbdosj-Input"
                  data-value=""
                >
                  <input
                    aria-activedescendant=""
                    aria-autocomplete="list"
                    aria-describedby="react-select-3-placeholder"
                    aria-expanded="false"
                    aria-haspopup="true"
                    aria-labelledby="label-ops-eval-and-learning-stakeholders"
                    autocapitalize="none"
                    autocomplete="off"
                    autocorrect="off"
                    class=""
                    id="react-select-3-input"
                    role="combobox"
                    spellcheck="false"
                    style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                    tabindex="0"
                    type="text"
                    value=""
                  />
                </div>
              </div>
              <div
                class=" css-1hb7zxy-IndicatorsContainer"
              >
                <span
                  class=" css-16b81-indicatorSeparator"
                />
                <div
                  aria-hidden="true"
                  class=" css-gjng1h-indicatorContainer"
                >
                  <svg
                    aria-hidden="true"
                    class="css-tj5bde-Svg"
                    focusable="false"
                    height="20"
                    viewBox="0 0 20 20"
                    width="20"
                  >
                    <path
                      d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                    />
                  </svg>
                </div>
              </div>
            </div>
            <div>
              <input
                name="stakeholders"
                type="hidden"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="margin-top-3 margin-bottom-6"
        >
          <button
            class="usa-button usa-button usa-button--unstyled"
            data-testid="add-note-toggle"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="usa-icon margin-right-1"
              focusable="false"
              height="1em"
              role="img"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
              />
            </svg>
            Add an additional note
          </button>
        </div>
      </div>
      <div
        class="usa-form-group margin-y-4 margin-bottom-8"
        data-scroll="ops-eval-and-learning-help-desk-use"
      >
        <label
          class="usa-label"
          data-testid="label"
          for="ops-eval-and-learning-help-desk-use"
        >
          Do you plan to use a helpdesk?
        </label>
        <fieldset
          class="usa-fieldset"
          data-testid="fieldset"
        >
          <div
            class="usa-radio"
            data-testid="radio"
          >
            <input
              class="usa-radio__input"
              data-testid="ops-eval-and-learning-help-desk-use-true"
              id="ops-eval-and-learning-help-desk-use-true"
              name="helpdeskUse"
              type="radio"
              value="TRUE"
            />
            <label
              class="usa-radio__label"
              for="ops-eval-and-learning-help-desk-use-true"
            >
              Yes
            </label>
          </div>
          <div
            class="usa-radio"
            data-testid="radio"
          >
            <input
              class="usa-radio__input"
              data-testid="ops-eval-and-learning-help-desk-use-false"
              id="ops-eval-and-learning-help-desk-use-false"
              name="helpdeskUse"
              type="radio"
              value="FALSE"
            />
            <label
              class="usa-radio__label"
              for="ops-eval-and-learning-help-desk-use-false"
            >
              No
            </label>
          </div>
        </fieldset>
        <div
          class="margin-top-3 margin-bottom-6"
        >
          <button
            class="usa-button usa-button usa-button--unstyled"
            data-testid="add-note-toggle"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="usa-icon margin-right-1"
              focusable="false"
              height="1em"
              role="img"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
              />
            </svg>
            Add an additional note
          </button>
        </div>
      </div>
      <div
        class="usa-form-group"
        data-scroll="ops-eval-and-learning-contractor-support"
      >
        <label
          class="usa-label"
          data-testid="label"
          for="ops-eval-and-learning-contractor-support"
        >
          What contractors will support your model?
        </label>
        <div
          class="mint-checkbox usa-checkbox"
        >
          <input
            class="usa-checkbox__input"
            id="ops-eval-and-learning-contractor-support-ONE"
            name="contractorSupport"
            type="checkbox"
            value="ONE"
          />
          <label
            class="usa-checkbox__label"
            for="ops-eval-and-learning-contractor-support-ONE"
          >
            One contractor to support implementation
          </label>
          <p
            class="margin-y-1 margin-left-4 line-height-body-3 font-body-2xs text-pre-line"
          />
        </div>
        <div
          class="mint-checkbox usa-checkbox"
        >
          <input
            class="usa-checkbox__input"
            id="ops-eval-and-learning-contractor-support-MULTIPLE"
            name="contractorSupport"
            type="checkbox"
            value="MULTIPLE"
          />
          <label
            class="usa-checkbox__label"
            for="ops-eval-and-learning-contractor-support-MULTIPLE"
          >
            May have separate contractors for different implementation functions
          </label>
          <p
            class="margin-y-1 margin-left-4 line-height-body-3 font-body-2xs text-pre-line"
          />
        </div>
        <div
          class="mint-checkbox usa-checkbox"
        >
          <input
            class="usa-checkbox__input"
            id="ops-eval-and-learning-contractor-support-NONE"
            name="contractorSupport"
            type="checkbox"
            value="NONE"
          />
          <label
            class="usa-checkbox__label"
            for="ops-eval-and-learning-contractor-support-NONE"
          >
            Do not plan to use an implemenation contractor
          </label>
          <p
            class="margin-y-1 margin-left-4 line-height-body-3 font-body-2xs text-pre-line"
          />
        </div>
        <div
          class="mint-checkbox usa-checkbox"
        >
          <input
            class="usa-checkbox__input"
            id="ops-eval-and-learning-contractor-support-OTHER"
            name="contractorSupport"
            type="checkbox"
            value="OTHER"
          />
          <label
            class="usa-checkbox__label"
            for="ops-eval-and-learning-contractor-support-OTHER"
          >
            Other
          </label>
          <p
            class="margin-y-1 margin-left-4 line-height-body-3 font-body-2xs text-pre-line"
          />
        </div>
        <div
          class="usa-form-group"
          data-scroll="ops-eval-and-learning-contractor-support-how"
        >
          <label
            class="usa-label text-normal margin-top-4"
            data-testid="label"
            for="ops-eval-and-learning-contractor-support-how"
          >
            In what capacity will they support your model?
          </label>
          <p
            class="text-base margin-y-1"
          >
            (implementation, data analysis, quality, etc.)
          </p>
          <textarea
            aria-labelledby="label-ops-eval-and-learning-contractor-support-how hint-ops-eval-and-learning-contractor-support-how"
            class="height-card usa-textarea"
            data-testid="ops-eval-and-learning-contractor-support-how"
            id="ops-eval-and-learning-contractor-support-how"
            name="contractorSupportHow"
          >
            Differ text
          </textarea>
        </div>
        <div
          class="margin-top-3 margin-bottom-6"
        >
          <button
            class="usa-button usa-button usa-button--unstyled"
            data-testid="add-note-toggle"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="usa-icon margin-right-1"
              focusable="false"
              height="1em"
              role="img"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
              />
            </svg>
            Add an additional note
          </button>
        </div>
      </div>
      <div
        class="usa-form-group margin-y-4 margin-bottom-8"
        data-scroll="ops-eval-and-learning-iddoc-support"
      >
        <label
          class="usa-label"
          data-testid="label"
          for="ops-eval-and-learning-iddoc-support"
        >
          Are you planning to use IDDOC support?
        </label>
        <p
          class="text-base margin-y-1"
        >
          IDDOC is commonly known as ACO-OS (Accountable Care Organization Operating System). They can provide support for design, development, operations, and maintenance.
        </p>
        <p
          class="text-base margin-y-1 margin-top-2"
        >
          If you select yes, there will be additional questions to answer.
        </p>
        <fieldset
          class="usa-fieldset"
          data-testid="fieldset"
        >
          <div
            class="usa-radio"
            data-testid="radio"
          >
            <input
              class="usa-radio__input"
              data-testid="ops-eval-and-learning-iddoc-support-true"
              id="ops-eval-and-learning-iddoc-support-true"
              name="iddocSupport"
              type="radio"
              value="TRUE"
            />
            <label
              class="usa-radio__label"
              for="ops-eval-and-learning-iddoc-support-true"
            >
              Yes
            </label>
          </div>
          <div
            class="usa-radio"
            data-testid="radio"
          >
            <input
              class="usa-radio__input"
              data-testid="ops-eval-and-learning-iddoc-support-false"
              id="ops-eval-and-learning-iddoc-support-false"
              name="iddocSupport"
              type="radio"
              value="FALSE"
            />
            <label
              class="usa-radio__label"
              for="ops-eval-and-learning-iddoc-support-false"
            >
              No
            </label>
          </div>
        </fieldset>
        <div
          class="margin-top-3 margin-bottom-6"
        >
          <button
            class="usa-button usa-button usa-button--unstyled"
            data-testid="add-note-toggle"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="usa-icon margin-right-1"
              focusable="false"
              height="1em"
              role="img"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
              />
            </svg>
            Add an additional note
          </button>
        </div>
      </div>
      <div
        class="margin-top-6 margin-bottom-3"
      >
        <button
          class="usa-button"
          data-testid="button"
          type="submit"
        >
          Next
        </button>
      </div>
      <button
        class="usa-button usa-button usa-button--unstyled"
        data-testid="button"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="usa-icon margin-right-1"
          focusable="false"
          height="1em"
          role="img"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
          />
        </svg>
        Save and return to task list
      </button>
    </fieldset>
  </form>
  <div
    class="mint-page-number margin-y-6"
  >
    <span
      class="mint-page-number__page-num"
    >
      Page 1 of 5
    </span>
  </div>
</DocumentFragment>
`;
